### ISSUE TITLE
Release 4.0

### ISSUE THREAD

**Issue by cfis** [jasonroelofs/rice#157]
I think its about time to do a 4.0 release, I have made the changes that I wanted to make. There is always more to do of course, but I think the main task left is updating the documentation (#152) and adding a migration section. And of course more testing is always helpful. I'd also like to add more stl support (std::map) but I don't think we should wait to do a release for that.

Anything I'm missing?

**Comment by jasonroelofs** [comment]
Works for me. I will try to spend some time this weekend getting the documentation set up in Sphinx, and when that's ready to go then shipping 4.0 sounds great!

**Comment by jasonroelofs** [comment]
We have two test failures yet on master, https://github.com/jasonroelofs/rice/runs/2303588167

Any thoughts on these? The windows one is weird, "bad any_cast", and the Ubuntu one may be GC related.

**Comment by cfis** [comment]
The windows one is weird, I looked at it the other day. I can duplicate it by running rake test but cannot duplicate it by running unitest.exe - although rake test runs unittest.exe first. It also weird its always Ruby 2.7 that fails.

The other one is caused by Ruby not freeing an object even through GC.stress is on.  Thus the destructor count is incorrect. It only happens intermittently so its mostly just an annoyance I think.

**Comment by jasonroelofs** [comment]
Alright docs change is merged into master. Feel free to add what you'd like to see and then we can get this release out!

**Comment by jasonroelofs** [comment]
I've narrowed down the `any_cast` failure to `pair.class_name()` (https://github.com/jasonroelofs/rice/runs/2333873971) though I'm not sure which of the three `any_cast` is failing (I should put some print outputs on those and see if I can track that down).

**Comment by jasonroelofs** [comment]
Ok and the `any_cast` in question is... MethodData!

https://github.com/jasonroelofs/rice/runs/2343486040?check_suite_focus=true

**Comment by cfis** [comment]
Ok, based on the changes in #160, everything I wanted to implement is done.  However, those changes do require updating the docs so I'll do that next. Another round of testing would be good. But then I think this is release is done.

**Comment by jasonroelofs** [comment]
Cool. We do have one test failure, thoughts on this?

```
Module:references: Assertion failed: boolValue != true (0 != 1) at test_Module.cpp:552
```

**Comment by cfis** [comment]
Yeah, that one is worrisome. I spent a bit of time trying to figure it out, but haven't yet. I do think it needs to be tracked down, will spend some more time on it.

**Comment by cfis** [comment]
Ok, that's fixed.

**Comment by jasonroelofs** [comment]
From #149 it sounds like we're in a good state? CI is all green, docs are looking good. Any further blockers to getting 4.0 out the door?

**Comment by cfis** [comment]
I'm good on my side for a release.

**Comment by cfis** [comment]
Anything you need from me @jasonroelofs?

**Comment by jasonroelofs** [comment]
@cfis Ah, no, I don't think so. I just need to make some time to finish this up.

Thanks again for all your work here, this has been awesome.

**Comment by cfis** [comment]
And thanks for being so open minded and collaborative!  Curious to see how this goes, I expect real world usage will reveal issues that need to get resolved....

**Comment by jasonroelofs** [comment]
Apologies for the delay, just pushed 4.0 to rubygems and I'll be crafting an announcement to send to Twitter and probably ruby-talk soon. 

cc @ankane for one final test run!

**Comment by ankane** [comment]
Nice! 🎉 

Re test run: it looks like `stl.hpp` is missing from the gemspec. https://github.com/jasonroelofs/rice/blob/c012811db999eff02631dafc9e802a8697f66211/rice.gemspec#L43

**Comment by ankane** [comment]
Also, not sure if you were planning to drop Ruby 2.5, but the current gemspec includes it.

**Comment by jasonroelofs** [comment]
Good catch re: `stl.hpp`. And yeah I'll drop 2.5 as well.

**Comment by jasonroelofs** [comment]
4.0.2 pushed with these fixes.

**Comment by ankane** [comment]
Awesome, working great now!

Would also be good to tag the releases. Bundler provides a `rake release` task to do this automatically if you include `require "bundler/gem_tasks"` in the `Rakefile`.

**Comment by cfis** [comment]
@jasonroelofs  - Nice work.

We should also submit an article to Peter Cooper at Ruby Weekly. We can do that here:

https://cooperpress.com/

You can do that or I can do that, let me know. And if I can find some time I can also write a blog post.

**Comment by jasonroelofs** [comment]
Right, been a while since I've done this! Tag created (and I should have -prereleased the initial versions to catch the kinks, oh well!)

**Comment by jasonroelofs** [comment]
Alright, announcements are out:

ruby-talk: http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-talk/441133
Twitter: https://twitter.com/jasonroelofs/status/1395010918082297861

@cfis If you'd like to write something up for Peter Cooper, feel free to do that!
